pkg/nsx/services/nsxserviceaccount/cluster.go

```diff
@@ -0,0 +1,289 @@
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/cache"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha1"
+	"github.com/vmware-tanzu/nsx-operator/pkg/logger"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+	"github.com/vmware-tanzu/nsx-operator/pkg/util"
+)
+
+const (
+	siteId             = "default"
+	enforcementpointId = "default"
+	PortRestAPI        = "rest-api"
+	PortNSXRPCFwdProxy = "nsx-rpc-fwd-proxy"
+	SecretSuffix       = "-nsx-cert"
+	SecretCAName       = "ca.crt"
+	SecretCertName     = "tls.crt"
+	SecretKeyName      = "tls.key"
+)
+
+var (
+	log = logger.Log
+
+	isProtectedTrue = true
+	vpcRole         = "ccp_internal_operator"
+	readerPath      = "/"
+	readerRole      = "cluster_info_reader"
+
+	antreaClusterResourceType = "AntreaClusterControlPlane"
+	revision1                 = int64(1)
+)
+
+type NSXServiceAccountService struct {
+	common.Service
+	PrincipalIdentityStore   *PrincipalIdentityStore
+	ClusterControlPlaneStore *ClusterControlPlaneStore
+}
+
+// InitializeNSXServiceAccount sync NSX resources
+func InitializeNSXServiceAccount(service common.Service) (*NSXServiceAccountService, error) {
+	wg := sync.WaitGroup{}
+	wgDone := make(chan bool)
+	fatalErrors := make(chan error)
+
+	wg.Add(2)
+	nsxServiceAccountService := &NSXServiceAccountService{Service: service}
+
+	nsxServiceAccountService.SetUpStore()
+	go nsxServiceAccountService.InitializeResourceStore(&wg, fatalErrors, common.ResourceTypePrincipalIdentity, nsxServiceAccountService.PrincipalIdentityStore)
+	go nsxServiceAccountService.InitializeResourceStore(&wg, fatalErrors, common.ResourceTypeClusterControlPlane, nsxServiceAccountService.ClusterControlPlaneStore)
+	go func() {
+		wg.Wait()
+		close(wgDone)
+	}()
+
+	select {
+	case <-wgDone:
+		break
+	case err := <-fatalErrors:
+		return nsxServiceAccountService, err
+	}
+
+	return nsxServiceAccountService, nil
+}
+
+func (s *NSXServiceAccountService) SetUpStore() {
+	s.PrincipalIdentityStore = &PrincipalIdentityStore{ResourceStore: common.ResourceStore{
+		Indexer:     cache.NewIndexer(keyFunc, cache.Indexers{common.TagScopeNSXServiceAccountCRUID: indexFunc}),
+		BindingType: mpmodel.PrincipalIdentityBindingType(),
+	}}
+	s.ClusterControlPlaneStore = &ClusterControlPlaneStore{ResourceStore: common.ResourceStore{
+		Indexer:     cache.NewIndexer(keyFunc, cache.Indexers{common.TagScopeNSXServiceAccountCRUID: indexFunc}),
+		BindingType: model.ClusterControlPlaneBindingType(),
+	}}
+}
+
+func (s *NSXServiceAccountService) CreateOrUpdateNSXServiceAccount(ctx context.Context, obj *v1alpha1.NSXServiceAccount) error {
+	clusterName := s.getClusterName(obj.Namespace, obj.Name)
+	normalizedClusterName := util.NormalizeId(clusterName)
+	// TODO: Support default project name in config as project
+	project := s.NSXConfig.CoeConfig.Cluster
+	vpcName := obj.Namespace
+	vpcPath := fmt.Sprintf("/orgs/default/projects/%s/vpcs/%s", util.NormalizeId(project), vpcName)
```

Comment:
for vpcPath, do we get from VPC CR status or joint path here?

Reply:
@lxiaopei @edwardbadboy 
Do we need to change the source of `project`/`vpcName`/`vpcPath`? Current implementation hardcoded them above.

Reply:
`project` is NSX cluster name.
`vpcName` is NSXServiceAccount CR namespace

Reply:
Hi @lxiaopei ,

Do you know how supervisor K8s admin specifies which VPC to use for workload cluster in an SV namespace?

If the workload cluster is using just the default VPC for the namespace, the Antrea-NSX integration should also use the default VPC in the namespace.

If supervisor cluster allows admin to specify different VPCs for different workload clusters in an SV namespace, Antrea-NSX integration should follow the same way.

In short, Antrea-NSX integration should always specify the same VPC as the workload cluster node VM is.

Reply:
Note:
We have `Spec.VPCName` in `NSXServiceAccount` CR

Reply:
for the first version there will be only 1 default VPC created for each namespace. 
for the default VPC name, it may not be the same as Namespace, might be "<namespace>-default-vpc".
As long as it will keep consistence with VPC implementation, it would be OK.

Reply:
OK. We need to fix here after the default VPC name is determined. Is this correct?

@edwardbadboy Do we need to check `Spec.VPCName` in `NSXServiceAccount` CR in this release?

Reply:
Do we finished the design for default VPC name?

Reply:
you can use "<namespace>-default-vpc" for now.

Reply:
Corrected default VPC name and added the check for `Spec.VPCName`

Reply:
Hi Ran,

If NCP and NSX-operator now support creating and reconciling new VPC CRs in a namespace, we can keep this Spec.VPCName. This VPCName should refer to the VPC CR name in the same namespace. If NCP and NSX-operator don't support custom VPCs (only support the namespace-default VPC), we can remove this VPCName field.

Reply:
Got it. Removed Spec.VPCName verification. This version we only support hard-coded default VPC per NS.